entrypoint.sh

```diff
@@ -1,136 +1,102 @@
-    fi
+	[[ "$is_default" = true ]] && array_contains "used_configs" "$config_name" && \
+		apply_value=false # Don't use default value if the user already set one!
+
+	[[ "$apply_value" != true ]] && return 0
+
+	echo "Setting config \"$config_name\" to: '$config_value'"
+	used_configs+=("$config_name")
+
+	# Append config to our on-the-fly-built config file
+	echo "${config_name}=${config_value}" >> "$CONFIG_FILE"
 }
-    
+
 # Drop the user into a shell, if they so wish
-if [ "$1" = "bash" ] || [ "$1" = "sh" ]; then
+if [[ "$1" = "bash" ||  "$1" = "sh" ]]; then
     echo "Dropping into interactive BASH session"
     exec "${@}"
 fi
 
-if [ -f "$MUMBLE_CUSTOM_CONFIG_FILE" ]; then
-	# Just use the config file specified by the user and don't bother assembling our own
+if [[ -f "$MUMBLE_CUSTOM_CONFIG_FILE" ]]; then
 	echo "Using manually specified config file at $MUMBLE_CUSTOM_CONFIG_FILE"
 	echo "All MUMBLE_CONFIG variables will be ignored"
 	CONFIG_FILE="$MUMBLE_CUSTOM_CONFIG_FILE"
 else
-	# As a first step, we ensure that the config file is empty, so we can always start from a clean slate
+	# Ensures the config file is empty, starting from a clean slate
 	echo -e "# Config file automatically generated from the MUMBLE_CONFIG_* environment variables\n" > "${CONFIG_FILE}"
 
-	used_configs=()
-	existing_config_options=()
-    
-	####
-	# Check what kind of configurations there exist in the bare bones config file
-	####
-	while read -r line; do
-		if [[ "$line" =~ ^(\;|\#)?\ *([a-zA-Z_0-9]+)=.* ]]; then
-			existing_config_options+=("${BASH_REMATCH[2]}")
-		fi
-	done < "$BARE_BONES_CONFIG_FILE"
-	####
-	# Process settings following environments variables starting with "MUMBLE_CONFIG_"
-	# Iterate over all environment variable key, value pairs and check if they match our naming scheme
-	####
+	# Process settings through variables of format MUMBLE_CONFIG_*
+
 	while IFS='=' read -d '' -r var value; do
-		if [[ "$var" == "MUMBLE_CONFIG_"* ]]; then
-			uppercase_variable=${var/MUMBLE_CONFIG_/}
-
-			# Remove underscores (to ensure that it doesn't matter whether the user
-			# uses e.g. MUMBLE_CONFIG_DB_BLA or MUMBLE_CONFIG_DBBLA)
-			uppercase_variable_no_underscores="${uppercase_variable//_/}"
-			found=false
-
-			for current_config in "${existing_config_options[@]}"; do
-				# convert to uppercase
-				upper_current_config=${current_config^^}
-				
-				if [ "$upper_current_config" = "$uppercase_variable" ] || [ "$upper_current_config" = "$uppercase_variable_no_underscores" ]; then
-					set_config "$current_config" "$value"
-					found=true
-					break
-				fi
-			done
-
-			if [ "$found" = "false" ]; then
-				>&2 echo "[ERROR]: Unable to find config corresponding to variable \"$var\""
-				exit 1
-			fi
+		# Uppercase and remove underscores so MUMBLE_CONFIG_A_B == MUMBLE_CONFIG_AB
+		uppercase_variable="${var/MUMBLE_CONFIG_/}"
+		config_option="${config_from_uppercase[${uppercase_variable//_/}]}"
```

Comment:
If I didn't miss anything, I think this right here introduces a possible bug. In case we ever introduced a config-option that does indeed contain an underscore in it, `config_from_uppercase` will also use the upper-cased config's name as a key. _Without_ having removed the underscore. Thus, by unconditionally removing the underscore from `uppercase_variable` _before_ the lookup, we would actually fail to find that config value.

Now, afaik we don't have any configs that use underscores yet, so I think the script should be working fine for all currently existing options. However, you never know what might be added in the future :)

That's why in the original script I first checked whether the upper-cased config name is a known option and only then check if the version with removed underscores is a known option.
That still doesn't cover the case where the original config option contains an underscore and the user added an additional one for readability, but at least when using the exact config name it'd work.

Another alternative would be to apply the same underscore-remove logic to the keys used in `config_from_uppercase` which has the underlying assumption that no two options differ only by casing and underscores. I guess that is a fine assumption though.

Reply:
Went with the latter assumption for simplicity in [3f908fd](https://github.com/mumble-voip/mumble-docker/pull/6/commits/3f908fd921883ef8c44ba61168864b501ff99f46). I hadn't realized it and hoped maybe INI files don't allow for keys with underscores, ridiculous as that sounds.

Thanks! :)